experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1376 @@
+    fn parse_exact_hex_digits<const N: usize>(&mut self) -> Result<[char; N]> {
+        let mut result = [0 as char; N];
+        for slot in result.iter_mut() {
+            let (offset, c) = self.must_next()?;
+            if !c.is_ascii_hexdigit() {
+                return Err(PEK::UnexpectedChar(c).with_offset(offset));
+            }
+            *slot = c;
+        }
+        Ok(result)
+    }
+
+    fn skip_whitespace(&mut self) {
+        while let Some(c) = self.peek_char() {
+            if !is_char_pattern_white_space(c) {
+                break;
+            }
+            self.iter.next();
+        }
+    }
+
+    fn consume(&mut self, expected: char) -> Result<()> {
+        match self.must_next()? {
+            (offset, c) if c != expected => Err(PEK::UnexpectedChar(c).with_offset(offset)),
+            _ => Ok(()),
+        }
+    }
+
+    // use this whenever an empty iterator would imply an Eof error
+    fn must_next(&mut self) -> Result<(usize, char)> {
+        self.iter.next().ok_or(PEK::Eof.into())
+    }
+
+    // use this whenever an empty iterator would imply an Eof error
+    fn must_peek(&mut self) -> Result<(usize, char)> {
+        self.iter.peek().copied().ok_or(PEK::Eof.into())
+    }
+
+    // see must_peek
+    fn must_peek_char(&mut self) -> Result<char> {
+        self.must_peek().map(|(_, c)| c)
+    }
+
+    // see must_peek
+    fn must_peek_index(&mut self) -> Result<usize> {
+        self.must_peek().map(|(idx, _)| idx)
+    }
+
+    fn peek_char(&mut self) -> Option<char> {
+        self.iter.peek().map(|&(_, c)| c)
+    }
+
+    // TODO: return Result<!> once ! is stable
+    #[inline]
+    fn error_here<T>(&mut self, kind: ParseErrorKind) -> Result<T> {
+        match self.iter.peek() {
+            None => Err(kind.into()),
+            Some(&(offset, _)) => Err(kind.with_offset(offset)),
+        }
+    }
+
+    fn process_strings(&mut self, op: Operation, other_strings: HashSet<String>) {
+        match op {
+            Operation::Union => self.multi_set.extend(other_strings.into_iter()),
+            Operation::Difference => {
+                self.multi_set = self.multi_set.difference(&other_strings).cloned().collect()
+            }
+            Operation::Intersection => {
+                self.multi_set = self
+                    .multi_set
+                    .intersection(&other_strings)
+                    .cloned()
+                    .collect()
+            }
+        }
+    }
+
+    fn process_chars(&mut self, op: Operation, other_chars: CodePointInversionList) {
+        match op {
+            Operation::Union => self.single_set.add_set(&other_chars),
+            Operation::Difference => self.single_set.remove_set(&other_chars),
+            Operation::Intersection => self.single_set.retain_set(&other_chars),
+        }
+    }
+
+    fn try_load_general_category_set(&mut self, name: &str) -> Result<()> {
+        // TODO(#3550): This could be cached; does not depend on name.
```

Comment:
Yes, we should; we should eagerly load everything we need when the parser is created, rather than caching a provider

We actually probably want to have a  UnicodeSetParserFactory (contains all the loaded name maps) and a UnicodeSetParser, contains a reference to the factory.

We probably shouldn't call it a factory though.

Reply:
Resolved in #3550 - trait + UnicodeSetParser struct reusable for multiple parses instead of factories